base_layer/core/src/mempool/utx_pool/utx_pool.rs

```diff
@@ -0,0 +1,316 @@
+#[derive(Clone, Copy)]
+pub struct UTxPoolConfig {
+    /// The maximum number of transactions that can be stored in the Unconfirmed Transaction pool
+    pub storage_capacity: usize,
+}
+
+impl Default for UTxPoolConfig {
+    fn default() -> Self {
+        Self {
+            storage_capacity: MEMPOOL_UTX_POOL_STORAGE_CAPACITY,
+        }
+    }
+}
+
+/// The Unconfirmed Transaction Pool consists of all unconfirmed transactions that are ready to be included in a block
+/// and they are prioritised according to the priority metric.
+pub struct UTxPool {
+    pool_storage: RwLock<UTxPoolStorage>,
+}
+
+impl UTxPool {
+    /// Create a new UTxPool with the specified configuration
+    pub fn new(config: UTxPoolConfig) -> Self {
+        Self {
+            pool_storage: RwLock::new(UTxPoolStorage::new(config)),
+        }
+    }
+
+    /// Insert a new transaction into the UTxPool. Low priority transactions will be removed to make space for higher
+    /// priority transactions. The lowest priority transactions will be removed when the maximum capacity is reached and
+    /// the new transaction has a higher priority than the currently stored lowest priority transaction.
+    pub fn insert_utx(&mut self, transaction: Transaction) -> Result<(), UTxPoolError> {
+        self.pool_storage
+            .write()
+            .map_err(|_| UTxPoolError::PoisonedAccess)?
+            .insert_utx(transaction)
+    }
+
+    ///  Insert a set of new transactions into the UTxPool
+    pub fn insert(&mut self, transactions: Vec<Transaction>) -> Result<(), UTxPoolError> {
+        self.pool_storage
+            .write()
+            .map_err(|_| UTxPoolError::PoisonedAccess)?
+            .insert(transactions)
+    }
+
+    /// Check if a transaction is available in the UTxPool
+    pub fn contains_key(&self, tx_excess_sig: &Signature) -> Result<bool, UTxPoolError> {
+        Ok(self
+            .pool_storage
+            .read()
+            .map_err(|_| UTxPoolError::PoisonedAccess)?
+            .contains_key(tx_excess_sig))
+    }
+
+    /// Returns a set of the highest priority unconfirmed transactions, that can be included in a block
+    pub fn highest_priority_utxs(&self, n: usize) -> Result<Vec<Transaction>, UTxPoolError> {
+        self.pool_storage
+            .read()
+            .map_err(|_| UTxPoolError::PoisonedAccess)?
+            .highest_priority_utxs(n)
+    }
+
+    /// Remove all published transactions from the UTxPool
+    pub fn remove_published(&mut self, published_block: &Block) -> Result<Vec<Transaction>, UTxPoolError> {
+        Ok(self
+            .pool_storage
+            .write()
+            .map_err(|_| UTxPoolError::PoisonedAccess)?
+            .remove_published(published_block))
+    }
+
+    /// Remove orphaned transactions from the UTxPool. These transactions were orphaned by the provided published block.
+    pub fn remove_orphaned(&mut self, published_block: &Block) -> Result<Vec<Transaction>, UTxPoolError> {
```

Comment:
I wonder if it's possible to do the removal of published and orphans at the same time. It seems like we could do one pass and return the published and orphaned at the same time. It also seems like the only time orphans would exist is after published txs have been removed

Reply:
Remove_published iterates over kernels and performs a hashmap lookup on utxs_by_signature. Remove_orphaned iterates over utxs_by_signature and inputs. As they are so different, I believe it will be inefficient to combine them. Also, separating them makes the code more understandable. The removal functions of some of the other sub-pools might be combinable, I will keep it in mind.

Reply:
This function was made private and it was combined to form: `remove_published_and_discard_double_spends`